src/steamship/agents/agents.py

```diff
@@ -0,0 +1,187 @@
+import uuid
+from abc import ABC, abstractmethod
+from typing import Any, List, Optional
+
+from pydantic import BaseModel
+
+from steamship import Block, File, Steamship, Tag, Task
+from steamship.invocable import InvocableResponse, PackageService, post
+from steamship.utils.kv_store import KeyValueStore
+
+
+class Action(BaseModel):
+    name: str
+    data: Any
+
+
+class FinishAction(Action):
+    pass
+
+
+class AgentContext(BaseModel, ABC):
+    # todo: define what should be in this context
+    # some set of state-related to agent (maybe request ids, callbacks, etc.)
+    pass
+
+    @abstractmethod
+    def update_blocks(self, blocks: List[Block]):
+        pass
+
+    @abstractmethod
+    def append_log(self, message: str):
+        pass
+
+
+class Tool(BaseModel, ABC):
+    name: str
+    human_description: str
+    ai_description: str
+
+    @abstractmethod
+    def run(self, tool_input: List[Block], context: AgentContext) -> List[Block]:
+        pass
+
+
+class KnockKnockTool(Tool):
+    name = "KnockKnockTool"
+    human_description = "Starts Knock-Knock Jokes"
+    ai_description = ("Used to begin the telling of a joke.",)
+
+    def run(self, tool_input: List[Block], context: AgentContext) -> List[Block]:
+        context.append_log("starting knock-knock joke...")
+        return [Block(text="Knock-Knock")]
+
+
+class WorkspaceTool(Tool, ABC):
+    client: Steamship
+
+
+class TaskScheduler(WorkspaceTool):
+    name = "TaskScheduler"
```

Comment:
Task Scheduler doesn't feel like a Tool to me, it feels like an aspect of when a tool would execute. Would love to find a good way to slide this laterally into a different part.